### PR TITLE
DS-3651: Range header support

### PR DESCRIPTION
**Pull request by tomdesair**
@Frederic-Atmire and I implemented a solution for https://jira.duraspace.org/browse/DS-3651 and https://jira.duraspace.org/browse/DS-3527 (for bitstreams).

Frederic created a SAF archive you can import and use to test this PR:
* https://www.dropbox.com/s/jsaoe9bd5o5biq2/DS-3651-SAF.zip?dl=0 (Note: this file will only remain available while the PR is open).
* Make sure the bitstream format registry is up to date: `bin/dspace registry-loader -bitstream config/registries/bitstream-formats.xml`
* `bin/dspace import -a -s ~/Downloads/ -z DS-3651-SAF.zip -c 123465789/2 -m mapfile -e admin@mail.com`

The SAF archive contains 3 items: one with a PDF, one with a MP4 file and one with a MP3 file. Each modern browser should open the files without problems. The MP4 and MP3 files should be downloaded by parts. You should also be able to jump within the media file to undownloaded sections.

Alternatively, you can also test using `curl` as illustrated in [the integration tests](https://github.com/DSpace/DSpace/pull/1884/files#diff-02eecb0d966d198c39e0d9400d7ac314R154) or with a Download mananger that can pause and resume downloads.

### CODE REVIEW

File: dspace-api/src/test/data/dspaceFolder/config/spring/api/solr-services.xml
Comment (abollini) [inline dspace-api/src/test/data/dspaceFolder/config/spring/api/solr-services.xml]: is this really required? the plan is to withdrawn ES support in DSpace7

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java]: please add the name of the main authors of the class

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java]: as said I'm unsure about adding the auth check here. In any case we should not deal with the response directly but throw a new REST runtime AuthorizeException that can be annotated see https://github.com/DSpace/DSpace/blob/master/dspace-spring-rest/src/main/java/org/dspace/app/rest/exception/RepositoryNotFoundException.java

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java]: are you sure about that?  an IOException can occur if we have issue to read from the underline storage. If the client abort the request the exception should occur on the Servlet container stack as writing to the response is in some way "buffered" by tomcat, etc. (I admit I haven't check)

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java]: No please, avoid catch all block. If we have other not runtime exceptions they should be encapsulated in a REST Exception to have single trasversal code to deal with logging and response code. We should move to the SpringMVC exception handling way https://spring.io/blog/2013/11/01/exception-handling-in-spring-mvc

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java]: please review the exception handing strategy according to springmvc

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java]: once you switch to the mvc exception handing this should be not required anymore

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/EPersonConverter.java
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/EPersonConverter.java]: I know that we have an open conversation about setting code style to avoid such kind of issue... but can you exclude from the PR the file that only have fix to the import? this should be done in a single dedicated pr when we have set our codestyle. This PR doesn't really need to touch more than 5-10 files I guess

File: dspace/config/dspace.cfg
Comment (abollini) [inline dspace/config/dspace.cfg]: can you add more details about the intend of the parameter? it is a limit for the range requests? it is discoverable in some way?

File: dspace/config/spring/rest/README.md
Comment (abollini) [inline dspace/config/spring/rest/README.md]: A good use case to place xml here is to allow override of the provided configuration such as the listners enabled for usage events that you have included

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java
Comment (abollini) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java]: We should throw the authorize exception and use the spring mvc excpetion handling here

**Review by abollini — CHANGES_REQUESTED** [review]

File: dspace-api/src/test/data/dspaceFolder/config/spring/api/solr-services.xml
Comment (tomdesair) [reply dspace-api/src/test/data/dspaceFolder/config/spring/api/solr-services.xml]: I moved these services from [here](https://github.com/DSpace/DSpace/blob/dspace-6.2/dspace/config/spring/api/core-services.xml#L95) to this new file so that we can override them in the integration tests.

I suggest we remove the `elasticSearchLoggerService` bean once we remove the complete Elastic Search implementation. I don't want to partly cleanup Elastic Search now with the risk of forgetting stuff.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java
Comment (tomdesair) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java]: We'll make the check more specific on the "Client aborted" use case.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java
Comment (tomdesair) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java]: Is it OK if we take the [global exception handling](https://spring.io/blog/2013/11/01/exception-handling-in-spring-mvc#global-exception-handling) approach using `@ControllerAdvice`?

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java
Comment (tomdesair) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java]: Ok

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java
Comment (tomdesair) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java]: We can also implement a `@ControllerAdvice` class that has a handler method for `AuthorizeException`s. That way we don't have to manually convert this exception to a REST runtime exception. What do you think?

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java
Comment (tomdesair) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java]: Our Range header utility class [also sets some error codes](https://github.com/atmire/DSpace/blob/ae0d6770f9e5ef03ec0d683fdcf1d6eab33e927b/dspace-spring-rest/src/main/java/org/dspace/app/rest/utils/MultipartFileSender.java#L254). So we cannot remove this method.

File: dspace/config/dspace.cfg
Comment (tomdesair) [reply dspace/config/dspace.cfg]: This is the size of the byte buffer [used for copying data from the bitstream FileInputStream to the request OutputStream](https://github.com/atmire/DSpace/blob/ae0d6770f9e5ef03ec0d683fdcf1d6eab33e927b/dspace-spring-rest/src/main/java/org/dspace/app/rest/utils/MultipartFileSender.java#L430). We can also omit the configuration property and give it a static value.

File: dspace/config/spring/rest/README.md
Comment (tomdesair) [reply dspace/config/spring/rest/README.md]: I hope that in the future we also find an easy way (using XML?) to override the default bean implementations we now have in dspace-spring-rest. For example, what if I want to use a custom `ItemConverter` bean in my own custom repository.

File: dspace-api/src/test/data/dspaceFolder/config/spring/api/solr-services.xml
Comment (abollini) [reply dspace-api/src/test/data/dspaceFolder/config/spring/api/solr-services.xml]: ok, I agree

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/configuration/RESTSpringLoader.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/configuration/RESTSpringLoader.java]: While we haven't really "policed" `@author` tags closely in recent years, I feel we should likely keep them pointing at individuals (who want credit for creating/refactoring specific classes) instead of organizations. I'm a bit worried that labeling classes with organizational contact info is a bit misleading, as it no longer refers to the specific creator of the class, and also could become an advertising gimmick/battle within our codebase (over which company gets more credit). The only company that "owns" the code is DuraSpace, and we just own the copyright in order to ensure the code remains open source, etc.

So, personally, I feel we should avoid this sort of change in our codebase. Either leave off the `@author` altogether, or reference the individual(s) who created / refactored the class.

However, if others disagree strongly here, I'll gladly bring this to discussion within the Committers group and/or Steering Group, so that we can formalize a policy on `@author` tags

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java]: While we haven't really "policed" `@author` tags closely in recent years, I feel we should likely keep them pointing at individuals (who want credit for creating/refactoring specific classes) instead of organizations. I'm a bit worried that labeling classes with organizational contact info is a bit misleading, as it no longer refers to the specific creator of the class, and also could become an advertising gimmick/battle within our codebase (over which company gets more credit). The only company that "owns" the code is DuraSpace, and we just own the copyright in order to ensure the code remains open source, etc.

So, personally, I feel we should avoid this sort of change in our codebase. Either leave off the `@author` altogether, or reference the individual(s) who created / refactored the class.

However, if others disagree strongly here, I'll gladly bring this to discussion within the Committers group and/or Steering Group, so that we can formalize a policy on `@author` tags

**Review by tdonohue — CHANGES_REQUESTED** [review]

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java
Comment (tdonohue) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java]: I only noted this in the first file I found with this company `@author` tag. There are a few others

Comment (tdonohue) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java]: @tomdesair : Thanks for pointing out the Lyncode `@author` tags. I suspect those slipped in during the massive OAI refactor (i.e. creation of XOAI) many years back. It's simply hard to police this sort of thing. Had I noticed these at the time, I also would have questioned them.  And, I'd rather remove them then add in more.

FWIW, other major projects (Apache Foundation and Gradle are two, see this [blog post for example](http://vojtechruzicka.com/stop-using-javadoc-author-tag/)) do not allow any `@author` tags whatsoever. Their policy seem to be they quickly get out of date or are misleading, and simply cause clutter.  While I understand your reasoning, I think individuals should be attributed via *commits* or via name, and not via a blanket generic statement.   As-is, I think "Atmire NV (info at atmire dot com)" gives us no information about who actually worked on this code at Atmire, so there's no way to attribute those individuals in our [Contributor list](https://wiki.duraspace.org/display/DSPACE/DSpaceContributors), or even (eventually) consider those individuals for Committership

**Comment by tomdesair** [comment]
@tdonohue I've corrected the author tags to unblock this PR. Once there is a decision on the usage of `@author` I'll update them accordingly in a separate PR.

**Review by tdonohue — APPROVED** [review]

**Review by abollini — APPROVED** [review]

**Comment by abollini** [comment]
I have solved a minor conflict in the dspace-spring-rest/src/test/java/org/dspace/app/rest/test/AbstractControllerIntegrationTest.java
and fixed dspace-spring-rest/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java#L120 (to switch to the static way to create builder for test as the constructor is now protected)

**Pull request merged**
